charms/istio-pilot/src/istioctl.py

```diff
@@ -0,0 +1,214 @@
+import logging
+import subprocess
+
+import yaml
+from charmed_kubeflow_chisme.lightkube.batch import delete_many
+from lightkube import Client, codecs
+
+
+class InstallFailedError(Exception):
+    pass
+
+
+class ManifestFailedError(Exception):
+    pass
+
+
+class PrecheckFailedError(Exception):
+    pass
+
+
+class UpgradeFailedError(Exception):
+    pass
+
+
+class VersionCheckError(Exception):
+    pass
+
+
+class Istioctl:
+    def __init__(
+        self, istioctl_path: str, namespace: str = "istio-system", profile: str = "minimal"
+    ):
+        """Wrapper for the istioctl binary
+
+        Args:
+            binary_file (str): Path to the istioctl binary to be used
+        """
+        self._istioctl_path = istioctl_path
+        self._namespace = namespace
+        self._profile = profile
+
+    @property
+    def _istioctl_flags(self):
+        return [
+            "-s",
+            f"profile={self._profile}",
+            "-s",
+            f"values.global.istioNamespace={self._namespace}",
+        ]
+
+    def install(self):
+        """Wrapper for the `istioctl install` command."""
+        try:
+            subprocess.check_call([self._istioctl_path, "install", "-y", *self._istioctl_flags])
+        except subprocess.CalledProcessError as cpe:
+            error_msg = f"Failed to install istio using istioctl.  Exit code: {cpe.returncode}."
+            logging.error(error_msg)
+            logging.error(f"stdout: {cpe.stdout}")
+            logging.error(f"stderr: {cpe.stderr}")
+
+            raise InstallFailedError(error_msg) from cpe
+
+    def manifest(self) -> str:
+        """Wrapper for the `istioctl manifest generate` command.
+
+        Returns:
+            (str) a YAML string of the Kubernetes manifest for Istio
+        """
+        try:
+            manifests = subprocess.check_output(
+                [self._istioctl_path, "manifest", "generate", *self._istioctl_flags]
+            )
+        except subprocess.CalledProcessError as cpe:
+            error_msg = (
+                f"Failed to generate manifests for istio using istioctl. "
+                f"Exit code: {cpe.returncode}."
+            )
+            logging.error(error_msg)
+            logging.error(f"stdout: {cpe.stdout}")
+            logging.error(f"stderr: {cpe.stderr}")
+
+            raise ManifestFailedError(error_msg) from cpe
+
+        return manifests
+
+    def precheck(self):
+        """Executes `istioctl x precheck` to validate whether the environment can be updated.
+
+        Raises:
+            PrecheckFailedError: if the precheck command fails.
+        """
```

Comment:
I think, we need a more info in comments regarding when `precheck` will succeed/fail. `precheck` will succeed on checking 1.11 against 1.16.2. And I understand that that's exactly what Istio precheck does, but since it is our wrapper, we can specify in comments or may be reference Istio doc that explains things.

Reply:
Yeah I would also like to see more of this, what would a successful precheck be and what does it mean, what useful information can we get from it?

Reply:
Suggested. Just fix my wording, if you'd like.

Reply:
lgtm!  ty